pkg/pci/pci.go

```diff
@@ -155,17 +155,11 @@ func (i *Info) String() string {
 // New returns a pointer to an Info struct that contains information about the
 // PCI devices on the host system
 func New(opts ...*option.Option) (*Info, error) {
-	return NewWithContext(context.New(opts...))
-}
-
-// NewWithContext returns a pointer to an Info struct that contains information about
-// the PCI devices on the host system. Use this function when you want to consume
-// the topology package from another package (e.g. gpu)
-func NewWithContext(ctx *context.Context) (*Info, error) {
+	ctx := context.New(opts...)
+	topo, err := topology.New(context.WithContext(ctx))
```

Comment:
from the consumer perspective the API looks nicer and more coherent with the rest of `ghw`, which is great. However here you'll still need to use a nested function (or some other trick) like I'm doing into
https://github.com/jaypipes/ghw/pull/278/commits/691659d5a0f68705b4ca7f3b4e285d22ad7a6cf2#diff-a6b7c579eae86e07718647003b278fce2940fccd215fa7db50f05f697691847fR158
to make sure that `topology.New` only runs *after* `pci` set the snapshot (if needed)

Reply:
ah, I see now. OK, fixing up.

Reply:
So I overlooked a key detail. Apologies! if you just add the trick I mentioned and run the tests, you'll get something like
```
-- FAIL: TestPCIDeviceRevision (0.22s)
    --- FAIL: TestPCIDeviceRevision/0000:07:03.0 (0.00s)
        pci_linux_test.go:86: got nil device for address "0000:07:03.0"
    --- FAIL: TestPCIDeviceRevision/0000:05:00.0 (0.00s)
        pci_linux_test.go:86: got nil device for address "0000:05:00.0"
setup(0xc00015e070) -> root="/tmp/ghw-snapshot-383180324"
setup(0xc00015e070) -> root="/tmp/ghw-snapshot-217854259"
teardown(0xc00015e070) -> root="/tmp/ghw-snapshot-217854259"
WARNING: failed to read /sys/bus/pci/devices
teardown(0xc00015e070) -> root="/tmp/ghw-snapshot-217854259"
```
(extra traces added by myself).

In the following of this post:
1. "outer package" = `pci`
2. "inner package" = `topology`

This means that this PR solves the problem of passing the context to subordinate packages, but still we get the snapshot handling incorrect:
1. outer package does setup
2. inner package does setup (again, wasteful but correct)
2. innerr package does their load
3. outer package *does teardown* - now the snapshot is gone but the state is inconsistent: the context passed through still believes the snapshot is availaible
4. outer package does their load - and this will fail!
5. outer package does teardown - which is a NOP at this point

IOW, the nesting of operations looks like
```
outer setup
   inner setup
     inner load
   inner teardown
   outer load
outer teardown
```

what we actually need is *only the outer package to do the setup/teardown* so we need a way to convey the inner packages to skip any setup/teardown. Just consume the context as is. It should look like
```
outer setup
    inner load
  outer load
outer teardown
```

So we need a way to convey the inner package "leave the context alone, just consume it". A minimal, and unfortunately a little uglier way, to do this could be:
1. in `context.go` add ```
2.

Reply:
Good detective work, @fromanirh :) I'll apply your patch and repush!